### DISCUSSION TITLE
validate xml in python code

### DISCUSSION THREAD

**Discussion by cudroiu** [CESNET/libyang#2157] [Q&A]
Hello Team,

I'm trying to use libyang to validate xml data coming from a device against a yang model. I'm using the sample code below. Is it possible to reveal all violations of the model present in the data? So far, I intentionally changed the data in 3 places (inventory-id, device-id and physical-port-id should have uuid format) so that it does not adhere to the model, but only the first violation is reported. Any chance to get information about all violations present in the data?
Thank you!
```python
ctx = libyang.Context('/var/opt/ncs')
mod = ctx.load_module('hw-inv.yang')

x1 = """
<hw-inv xmlns="urn:ietf:params:xml:ns:yang:hw-inv">
    <inventory>
        <inventory-id>b19e2b2a</inventory-id>
        <device>
            <device-id>039f552f</device-id>
            <physical-port>
                <physical-port-id>0a0c</physical-port-id>
                <physical-port-name>portB/5</physical-port-name>
                <parent-component xmlns="http://openconfig.net/yang/platform">port 1/1/20</parent-component>
            </physical-port>
        </device>
    </inventory>
</hw-inv>"""
dnode = ctx.parse_data_mem(x1, 'xml', validate_present=True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/nsoadmin/.local/lib/python3.10/site-packages/libyang/context.py", line 393, in parse_data_mem
    return self.parse_data(
  File "/home/nsoadmin/.local/lib/python3.10/site-packages/libyang/context.py", line 373, in parse_data
    raise self.error("failed to parse data tree")
libyang.util.LibyangError: failed to parse data tree: Unsatisfied pattern - "b19e2b2a" does not conform to "[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}".
```

**Comment by michalvasko** [comment]
Yes, try using the `LYD_VALIDATE_MULTI_ERROR` validation flag but whether it is available in the Python bindings I do not know.

**Reply by cudroiu** [reply]
Thank you Michal!
Do you have an example of using this flag?

**Reply by michalvasko** [reply]
I do not understand what would you need an example for, just use the flag when calling `parse_data_mem` in your case, in the validation flags. But `yanglint` uses it when you use the command `data` or simply parse a data file in the non-interactive version.

**Reply by cudroiu** [reply]
Thank you Michal!
I was asking because I don't see an argument in parse_data_mem where we can provide a list of flags. Below is output from the help method on parse_data_mem:

`parse_data_mem(data: Union[str, bytes], fmt: str, parent: libyang.data.DNode = None, lyb_mod_update: bool = False, no_state: bool = False, parse_only: bool = False, opaq: bool = False, ordered: bool = False, strict: bool = False, validate_present: bool = False) -> Optional[libyang.data.DNode] method of libyang.context.Context instance`

**Reply by michalvasko** [reply]
Seems that each flag was transformed into a separate parameter but this one is missing so to use it the bindings would need to be updated.

**Reply by cudroiu** [reply]
Thanks Michal!
I'll try to see how I can add these bindings.

**Reply by cudroiu** [reply]
Hi Michal,

I don't think I understand what you mean by "bindings would need to be updated"?

**Reply by jktjkt** [reply]
You need to talk to people who maintain libyang-python to update their code so that this feature is made available.

**Reply by cudroiu** [reply]
Thanks a lot!
